File: frontend/lib/model/judgement/usepairjudgement/model/UsePairJudgement.ts
import UsePairInstance from "../../../instance/usepairinstance/model/UsePairInstance";
import { IJudgement, IJudgementConstructor } from "../../model/IJudgement";
import UsePairJudgementDto from "../dto/UsePairJudgementDto";
import UsePairJudgementId from "./UsePairJudgementId";

export default class UsePairJudgement implements IJudgement {

    readonly id: UsePairJudgementId;

    readonly instance: UsePairInstance;
    
    readonly label: string;
    readonly comment: string;

    constructor(id: UsePairJudgementId, instance: UsePairInstance, label: string, comment: string) {
        this.id = id;
        this.instance = instance;
        this.label = label;
        this.comment = comment;
    }

    public getId(): UsePairJudgementId {
        return this.id;
    }

    public getInstance(): UsePairInstance {
        return this.instance;
    }

    public getLabel(): string {
        return this.label;
    }

    public getComment(): string {
        return this.comment;
    }

    public static fromDto(dto: UsePairJudgementDto): UsePairJudgement {
        return new UsePairJudgement(
            UsePairJudgementId.fromDto(dto.id),
            UsePairInstance.fromDto(dto.instance),
            dto.label,
            dto.comment
        );
    }

}

export class UsePairJudgementConstructor implements IJudgementConstructor {
    fromDto(dto: UsePairJudgementDto): UsePairJudgement {
        return UsePairJudgement.fromDto(dto);
    }
}

File: backend/src/main/java/de/garrafao/phitag/domain/judgement/usepairjudgement/UsePairJudgement.java
package de.garrafao.phitag.domain.judgement.usepairjudgement;

import de.garrafao.phitag.domain.annotator.Annotator;
import de.garrafao.phitag.domain.instance.usepairinstance.UsePairInstance;
import de.garrafao.phitag.domain.judgement.IJudgement;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.Validate;

import javax.persistence.*;

@Entity
@Table(name = "phitagusepairjudgement")
@Getter
public class UsePairJudgement implements IJudgement {

    @Setter
    @EmbeddedId
    private UsePairJudgementId id;

    @MapsId("annotatorid")
    @ManyToOne
    private Annotator annotator;

    @MapsId("instanceid")
    @ManyToOne
    private UsePairInstance instance;

    @Setter
    @Column(name = "label", nullable = false)
    private String label;

    @Setter
    @Column(name = "comment", nullable = false)
    private String comment;

    public UsePairJudgement() {
    }

    public UsePairJudgement(UsePairInstance usePairInstance, Annotator annotator, String label, String comment) {
        Validate.notNull(usePairInstance, "instanceData must not be null");
        Validate.notNull(annotator, "annotator must not be null");
        Validate.notNull(label, "label must not be null");
        Validate.notNull(comment, "comment must not be null");

        this.id = new UsePairJudgementId(usePairInstance.getId(), annotator.getId());

        this.instance = usePairInstance;
        this.annotator = annotator;

        this.label = label;
        this.comment = comment;
    }

    public UsePairInstance getUsePairInstance() {
        return instance;
    }

    @Override
    public String toString() {
        return "UsePairJudgement [id=" + id + ", label=" + label + ", comment=" + comment + "]";
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        UsePairJudgement other = (UsePairJudgement) obj;
        return this.id.equals(other.getId());
    }

    @Override
    public int hashCode() {
        return this.id.hashCode();
    }

}
